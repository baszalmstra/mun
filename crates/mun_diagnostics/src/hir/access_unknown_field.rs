use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_hir::HirDisplay;
use mun_syntax::{ast, AstNode, TextRange};

/// An error that is emitted when trying to access a field that doesn't exist.
///
/// ```mun
/// struct Foo {
///     b: i32
/// }
///
/// # fn main() {
/// let a = Foo { b: 3}
/// let b = a.c;    // no field `c`
/// #}
/// ```
pub struct AccessUnknownField<'db, 'diag> {
    db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::AccessUnknownField,
    location: TextRange,
}

impl<'db, 'diag> Diagnostic for AccessUnknownField<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.location
    }

    fn title(&self) -> String {
        format!(
            "no field `{}` on type `{}`",
            self.diag.name,
            self.diag.receiver_ty.display(self.db),
        )
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.location,
            message: "unknown field".to_string(),
        })
    }
}

impl<'db, 'diag> AccessUnknownField<'db, 'diag> {
    /// Constructs a new instance of `AccessUnknownField`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::AccessUnknownField,
    ) -> Self {
        let parse = db.parse(diag.file);

        let location = ast::FieldExpr::cast(diag.expr.to_node(&parse.syntax_node()))
            .map(|f| f.field_range())
            .unwrap_or_else(|| diag.highlight_range());

        AccessUnknownField { db, diag, location }
    }
}
