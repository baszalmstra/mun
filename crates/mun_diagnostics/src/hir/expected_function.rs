use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_hir::HirDisplay;
use mun_syntax::TextRange;

/// An error that is emitted when a function is expected but something else is encountered:
///
/// ```mun
/// # fn main() {
///     let a = 3;
///     let b = a();    // expected function
/// # }
/// ```
pub struct ExpectedFunction<'db, 'diag> {
    db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::ExpectedFunction,
}

impl<'db, 'diag> Diagnostic for ExpectedFunction<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!(
            "expected function, found `{}`",
            self.diag.found.display(self.db)
        )
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.diag.highlight_range(),
            message: "not a function".to_owned(),
        })
    }
}

impl<'db, 'diag> ExpectedFunction<'db, 'diag> {
    /// Constructs a new instance of `ExpectedFunction`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::ExpectedFunction,
    ) -> Self {
        ExpectedFunction { db, diag }
    }
}
