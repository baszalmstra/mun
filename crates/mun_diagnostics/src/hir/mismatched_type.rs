use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_hir::HirDisplay;
use mun_syntax::TextRange;

/// An error that is emitted when a different type was found than expected.
///
/// ```mun
/// fn add(a: i32, b: i32) -> i32{
///     a+b
/// }
///
/// # fn main() {
///     add(true, false); // type mismatch, expected i32 found bool.
/// # }
/// ```
pub struct MismatchedType<'db, 'diag> {
    db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::MismatchedType,
}

impl<'db, 'diag> Diagnostic for MismatchedType<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!(
            "expected `{}`, found `{}`",
            self.diag.expected.display(self.db),
            self.diag.found.display(self.db)
        )
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        None
    }
}

impl<'db, 'diag> MismatchedType<'db, 'diag> {
    /// Constructs a new instance of `MismatchedType`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::MismatchedType,
    ) -> Self {
        MismatchedType { db, diag }
    }
}
