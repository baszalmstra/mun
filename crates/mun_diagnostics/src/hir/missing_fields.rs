use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_hir::HirDisplay;
use mun_syntax::{ast, AstNode, TextRange};

/// An error that is emitted when a field is missing from a struct initializer.
///
/// ```mun
/// struct Foo {
///     a: i32,
/// }
///
/// # fn main() {
///     let a = Foo {}; // missing field `a`
/// # }
/// ```
pub struct MissingFields<'db, 'diag> {
    db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::MissingFields,
    location: TextRange,
    missing_fields: String,
}

impl<'db, 'diag> Diagnostic for MissingFields<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.location
    }

    fn title(&self) -> String {
        format!(
            "missing fields {} in initializer of `{}`",
            self.missing_fields,
            self.diag.struct_ty.display(self.db)
        )
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.location,
            message: format!("missing {}", self.missing_fields.clone()),
        })
    }
}

impl<'db, 'diag> MissingFields<'db, 'diag> {
    /// Constructs a new instance of `MissingFields`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::MissingFields,
    ) -> Self {
        let parse = db.parse(diag.file);
        let missing_fields = diag
            .field_names
            .iter()
            .map(|n| format!("`{}`", n))
            .collect::<Vec<String>>()
            .join(", ");

        let location = ast::RecordLit::cast(diag.fields.to_node(&parse.syntax_node()))
            .and_then(|f| f.type_ref())
            .map(|t| t.syntax().text_range())
            .unwrap_or_else(|| diag.highlight_range());

        MissingFields {
            db,
            diag,
            location,
            missing_fields,
        }
    }
}
