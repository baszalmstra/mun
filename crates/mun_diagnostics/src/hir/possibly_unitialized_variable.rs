use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_syntax::TextRange;

/// An error that is emitted when trying to access a field that is potentially not yet initialized.
///
/// ```mun
/// # fn main() {
/// let a;
/// let b = a;    // `a` is possible not yet initialized
/// #}
/// ```
pub struct PossiblyUninitializedVariable<'db, 'diag> {
    _db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::PossiblyUninitializedVariable,
    value_name: String,
}

impl<'db, 'diag> Diagnostic for PossiblyUninitializedVariable<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!("use of possibly-uninitialized `{}`", self.value_name)
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        None
    }
}

impl<'db, 'diag> PossiblyUninitializedVariable<'db, 'diag> {
    /// Constructs a new instance of `PossiblyUninitializedVariable`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::PossiblyUninitializedVariable,
    ) -> Self {
        let parse = db.parse(diag.file);

        // Get the text of the value as a string
        let value_name = diag.pat.to_node(&parse.syntax_node()).text().to_string();

        PossiblyUninitializedVariable {
            _db: db,
            diag,
            value_name,
        }
    }
}
