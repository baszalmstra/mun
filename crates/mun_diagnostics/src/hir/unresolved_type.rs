use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_syntax::{AstNode, TextRange};

/// An error that is emitted when trying to use a type that doesnt exist within the scope.
///
/// ```mun
/// # fn main() {
/// let a = DoesntExist {}; // Cannot find `DoesntExist` in this scope.
/// #}
/// ```
pub struct UnresolvedType<'db, 'diag> {
    _db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::UnresolvedType,
    value_name: String,
}

impl<'db, 'diag> Diagnostic for UnresolvedType<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!("cannot find type `{}` in this scope", self.value_name)
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.diag.highlight_range(),
            message: "not found in this scope".to_owned(),
        })
    }
}

impl<'db, 'diag> UnresolvedType<'db, 'diag> {
    /// Constructs a new instance of `UnresolvedType`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::UnresolvedType,
    ) -> Self {
        let parse = db.parse(diag.file);

        // Get the text of the value as a string
        let value_name = diag
            .type_ref
            .to_node(&parse.syntax_node())
            .syntax()
            .text()
            .to_string();

        UnresolvedType {
            _db: db,
            diag,
            value_name,
        }
    }
}
