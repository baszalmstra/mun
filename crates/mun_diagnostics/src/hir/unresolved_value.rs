use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_syntax::{AstNode, TextRange};

/// An error that is emitted when trying to use a value that doesnt exist within the scope.
///
/// ```mun
/// # fn main() {
/// let a = b; // Cannot find `b` in this scope.
/// #}
/// ```
pub struct UnresolvedValue<'db, 'diag> {
    _db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::UnresolvedValue,
    value_name: String,
}

impl<'db, 'diag> Diagnostic for UnresolvedValue<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!("cannot find value `{}` in this scope", self.value_name)
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.diag.highlight_range(),
            message: "not found in this scope".to_owned(),
        })
    }
}

impl<'db, 'diag> UnresolvedValue<'db, 'diag> {
    /// Constructs a new instance of `UnresolvedValue`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::UnresolvedValue,
    ) -> Self {
        let parse = db.parse(diag.file);

        // Get the text of the value as a string
        let value_name = diag.expr.to_node(&parse.tree().syntax()).text().to_string();

        UnresolvedValue {
            _db: db,
            diag,
            value_name,
        }
    }
}
