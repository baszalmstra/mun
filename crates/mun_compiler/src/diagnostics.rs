use crate::diagnostics_snippets::{emit_hir_diagnostic, emit_syntax_error};
use crate::DisplayColor;
use hir::{DiagnosticSink, HirDatabase};
use std::io::Cursor;

/// Emits all diagnostic messages currently in the database; returns true if errors were
/// emitted.
pub fn emit_diagnostics(
    db: &dyn HirDatabase,
    writer: &mut dyn std::io::Write,
    display_color: DisplayColor,
) -> Result<bool, anyhow::Error> {
    let emit_colors = display_color.should_enable();
    let mut has_error = false;

    for package in hir::Package::all(db) {
        for module in package.modules(db) {
            if let Some(file_id) = module.file_id(db) {
                let parse = db.parse(file_id);
                let source_code = db.file_text(file_id);
                let relative_file_path = db.file_relative_path(file_id);
                let line_index = db.line_index(file_id);

                // Emit all syntax diagnostics
                for syntax_error in parse.errors().iter() {
                    emit_syntax_error(
                        syntax_error,
                        relative_file_path.as_str(),
                        &source_code,
                        &line_index,
                        emit_colors,
                        writer,
                    )?;
                    has_error = true;
                }

                // Emit all HIR diagnostics
                let mut error = None;
                module.diagnostics(
                    db,
                    &mut DiagnosticSink::new(|d| {
                        has_error = true;
                        if let Err(e) = emit_hir_diagnostic(d, db, file_id, emit_colors, writer) {
                            error = Some(e)
                        };
                    }),
                );

                // If an error occurred when emitting HIR diagnostics, return early with the error.
                if let Some(e) = error {
                    return Err(e.into());
                }
            }
        }
    }

    Ok(has_error)
}

/// Returns all diagnostics as a human readable string
pub fn emit_diagnostics_to_string(
    db: &dyn HirDatabase,
    display_color: DisplayColor,
) -> anyhow::Result<Option<String>> {
    let mut compiler_errors: Vec<u8> = Vec::new();
    if !emit_diagnostics(db, &mut Cursor::new(&mut compiler_errors), display_color)? {
        Ok(None)
    } else {
        Ok(Some(String::from_utf8(compiler_errors).map_err(|e| {
            anyhow::anyhow!(
                "could not convert compiler diagnostics to valid UTF8: {}",
                e
            )
        })?))
    }
}

#[cfg(test)]
mod tests {
    use super::emit_diagnostics_to_string;
    use crate::{Config, DisplayColor, Driver, PathOrInline, RelativePathBuf};

    /// Compile passed source code and return all compilation errors
    fn compilation_errors(source_code: &str) -> String {
        let config = Config::default();

        let input = PathOrInline::Inline {
            rel_path: RelativePathBuf::from("main.mun"),
            contents: source_code.to_owned(),
        };

        let (driver, _) = Driver::with_file(config, input).unwrap();
        emit_diagnostics_to_string(driver.database(), DisplayColor::Disable)
            .unwrap()
            .unwrap_or_default()
    }

    #[test]
    fn test_syntax_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\nfn main(\n struct Foo\n"));
    }

    #[test]
    fn test_unresolved_value_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet b = a;\n\nlet d = c;\n}"
        ));
    }

    #[test]
    fn test_unresolved_type_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a = Foo{};\n\nlet b = Bar{};\n}"
        ));
    }

    #[test]
    fn test_expected_function_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a = Foo();\n\nlet b = Bar();\n}"
        ));
    }

    #[test]
    fn test_mismatched_type_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a: f64 = false;\n\nlet b: bool = 22;\n}"
        ));
    }

    #[test]
    fn test_duplicate_definition_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn foo(){}\n\nfn foo(){}\n\nstruct Bar;\n\nstruct Bar;\n\nfn BAZ(){}\n\nstruct BAZ;"
        ));
    }

    #[test]
    fn test_possibly_uninitialized_variable_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a;\nif 5>6 {\na = 5\n}\nlet b = a;\n}"
        ));
    }

    #[test]
    fn test_access_unknown_field_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nstruct Foo {\ni: bool\n}\n\nfn main() {\nlet a = Foo { i: false };\nlet b = a.t;\n}"
        ));
    }

    #[test]
    fn test_free_type_alias_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\ntype Foo;"));
    }

    #[test]
    fn test_type_alias_target_undeclared_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\ntype Foo = UnknownType;"));
    }

    #[test]
    fn test_cyclic_type_alias_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\ntype Foo = Foo;"));
    }

    #[test]
    fn test_expected_function() {
        insta::assert_display_snapshot!(compilation_errors("\n\nfn foo() { let a = 3; a(); }"));
    }
}
