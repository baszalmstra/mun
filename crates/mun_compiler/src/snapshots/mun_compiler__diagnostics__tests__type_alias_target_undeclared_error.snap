---
source: crates/mun_compiler/src/diagnostics.rs
assertion_line: 97
expression: "compilation_errors(\"\\n\\ntype Foo = UnknownType;\")"

---
error: cannot find type `UnknownType` in this scope
 --> main.mun:3:12
  |
3 | type Foo = UnknownType;
  |            ^^^^^^^^^^^ not found in this scope
  |
